### DISCUSSION TITLE
iface_ctx.off check and modem reset high/low time in HL7800 driver

### DISCUSSION THREAD

**Discussion by pj-sunny** [zephyrproject-rtos/zephyr#66004] [Q&A] [answered]
Hi,

For the `iface_ctx.off` check introduced in the commit [here](https://github.com/zephyrproject-rtos/zephyr/commit/dd195d6ea5b9135bc2bdc446a50bd8d1c4d5692d), the error message always shows for me: `modem_boot_handler: Err waiting for boot: -11`.
It turns out that although the `mdm_vgpio_callback_isr()` is called within 1ms, it does nothing because the `iface_ctx.off` is still `true`.
And this is because the `iface_ctx.off` will not be set to `false` until the 10ms `MDM_RESET_HIGH_TIME` in `modem_run()` passed.

This seems to be an issue to me because the delay between RESET_IN_N and VGPIO is maximum 1ms according to the HL7800 product technical specification (Table 4-10).
In my opinion, the `iface_ctx.off` should be set to `false` right after the `gpio_pin_set_dt()` instead of after the 10ms `MDM_RESET_HIGH_TIME` because the driver has requested the modem to run by setting the reset pin high.

But not sure why no one else raises similar question.
Maybe this is just my question.

Also the specification defines the minimum reset low time 100us and has no definition about the reset high time.
Actually I do not understand why there is 50ms `MDM_RESET_LOW_TIME` with comment `/* >20 milliseconds required for reset low */` and 10ms `MDM_RESET_HIGH_TIME`.
Of course they are necessary for implementation but seem to be a little bit longer than needed.

Maybe there are some reasons for these implementations and hope someone can help to share some insights.
Thank you.

**Comment by zafersn** [comment]
Hi @pj-sunny 

> For the iface_ctx.off check introduced in the commit [here](https://github.com/zephyrproject-rtos/zephyr/commit/dd195d6ea5b9135bc2bdc446a50bd8d1c4d5692d), the error message always shows for me: modem_boot_handler: Err waiting for boot: -11.

do you have the pm_device enabled?

I had the same issue `modem_boot_handler: Err waiting for boot: -11` when  I have `CONFIG_PM_DEVICE=y` enabled 

the issue was that forgotten uart enabling.

see `shutdown_uart()` function in `prepare_io_for_reset ` in  `modem_reset`  function. here they shut down the uart but after that.
I think during the `modem_run ` they should've enabled it too, But i need to add by myself and worked.


P.S. 
I think there are more conflicts in this driver file. between PM_ENABLE and LOW_POWER_MODE

**Reply by pj-sunny** [reply]
Hi @zafersn ,

Thank you for the response.
Yes, I do have the `CONFIG_PM_DEVICE` enabled.
But in my opinion, the `CONFIG_PM_DEVICE` is not the root cause.

Also the flow to enable the uart, i.e., call the `power_on_uart()`, is in `mdm_vgpio_callback_isr()`, which is not executed because the `iface_ctx.off` is still true.
Before the `iface_ctx.off` check introduced, actually everything worked correctly on my side.

**Reply by zafersn** [reply]
@pj-sunny I agree with what you stated there, If anyone wants a quick start, then they basically need to disable `CONFIG_PM_DEVICE ` and then can continue. Just a heads up. Otherwise, they need to consider all the changes you explained in the question very well.

**Comment by pj-sunny** [comment]
It seems to me that the issue is resolved by [PR 66237](https://github.com/zephyrproject-rtos/zephyr/pull/66237).
@rerickson1 , would you please help to share some insights about the implementations?
Thank you.

![image](https://github.com/zephyrproject-rtos/zephyr/assets/152298697/e4c5a5f7-c235-4fb2-bc82-6588436bd7a6)

**Comment by rerickson1** [comment, marked as answer]
> here

Hello, yes, that PR should have solve all issues regarding this.
